pyccel/codegen/compiling/project.py

```diff
@@ -0,0 +1,327 @@
+        if deps:
+            cycle = [next(c for c in deps)]
+            while len(cycle) < 2 or cycle[-1] != cycle[0]:
+                c = cycle[-1]
+                unfulfilled_dep = next(d for d in deps[c] if d not in placed)
+                cycle.append(next(c for c in deps if getattr(c, 'pyfile', getattr(c, 'folder')) == unfulfilled_dep))
+
+            cycle_example = ' -> '.join(str(getattr(c, 'pyfile', c.folder)) for c in cycle)
+
+            errors.report(f"Found circular dependencies between directories: {cycle_example}",
+                         severity='fatal')
+
+        self._folder = folder
+        self._targets = targets
+        self._dependencies = {d for t in self._targets for d in t.dependencies if d not in self}
+
+    @property
+    def dependencies(self):
+        """
+        Get all directories which must be compiled before this directory.
+
+        Get all directories which must be compiled before this directory.
+        """
+        return self._dependencies
+
+    @property
+    def folder(self):
+        """
+        Get the path to the folder being described by this target.
+
+        Get the path to the folder being described by this target.
+        """
+        return self._folder
+
+    @property
+    def targets(self):
+        """
+        Get all targets found in this directory.
+
+        Get all targets found in this directory. This includes compilation targets
+        and sub-directories.
+        """
+        return self._targets
+
+    def __contains__(self, other):
+        if isinstance(other, CompileTarget):
+            return self.folder in other.pyfile.parents
+        else:
+            return self.folder in other.folder.parents
+
+    def __repr__(self):
+        return f'DirTarget({self.folder})'
+
+class BuildProject:
+    """
+    Class representing the overall build project structure.
+
+    This class encapsulates the directory structure, compilation targets,
+    programming languages, and standard library dependencies of a project.
+    It serves as the main data container for build configuration.
+
+    Parameters
+    ----------
+    root_dir : str | Path
+        Root directory of the project where the original Python code is found.
+    compile_targets : iterable[CompileTarget]
+        An iterable of all compile targets in the project.
+    languages : iterable[str]
+        An iterable of languages used in the project (e.g., ['C', 'Fortran']).
+    stdlib_deps : dict[str, CompileObj]
+        A dictionary mapping the names of standard library dependencies
+        required for the build to the CompileObj describing how they are used.
+    """
+    def __init__(self, root_dir, compile_targets, languages, stdlib_deps):
+        self._root_dir = Path(root_dir)
+
+        self._dir_info = DirTarget(self._root_dir, compile_targets)
+
+        self._languages = languages
+
+        self._stdlib_deps = stdlib_deps
```

Comment:
What is the purpose of the empty lines?

Reply:
```suggestion
    def __init__(self, root_dir, compile_targets, languages, stdlib_deps):
        self._root_dir = Path(root_dir)
        self._dir_info = DirTarget(self._root_dir, compile_targets)
        self._languages = languages
        self._stdlib_deps = stdlib_deps
```

Reply:
Done : [d498e93](https://github.com/pyccel/pyccel/pull/2383/commits/d498e93f58dc082b795c1b20b0a77ae739b95ed6)
I fell asleep before pushing it XD

Reply:
A bit of jet lag is unavoidable! ✈️